rpc/dialer.go

```diff
@@ -240,6 +241,152 @@ func DialDirectGRPC(ctx context.Context, address string, logger utils.ZapCompati
+		return nil
+	}
+
+	return func(ctx context.Context, network, addr string) (net.Conn, error) {
+		// if ONLY_SOCKS_PROXY specified, no need for a parallel dial - only dial through
+		// the SOCKS proxy directly.
+		if os.Getenv(OnlySocksProxyEnvVar) != "" {
+			logger.Infow("Both SOCKS_PROXY and ONLY_SOCKS_PROXY specified, only SOCKS proxy will be used for outgoing connection")
+			conn, err := socksProxyDialContext(ctx, network, proxyAddr, addr)
+			if err == nil {
+				logger.Infow("connected with SOCKS proxy")
+			}
+			return conn, err
+		}
+
+		type dialResult struct {
+			net.Conn
+			error
+			primary bool
+			done    bool
+		}
+		results := make(chan dialResult) // unbuffered
+
+		var primary, fallback dialResult
+		var wg sync.WaitGroup
+		defer wg.Wait()
+
+		// otherwise, do a parallel dial with a slight delay for the fallback option.
+		returned := make(chan struct{})
+		defer close(returned)
+
+		dialer := func(ctx context.Context, dialFunc func(context.Context) (net.Conn, error), primary bool) {
+			defer wg.Done()
+			conn, err := dialFunc(ctx)
+			select {
+			case results <- dialResult{Conn: conn, error: err, primary: primary, done: true}:
+			case <-returned:
+				if conn != nil {
+					utils.UncheckedError(conn.Close())
+				}
+			}
+		}
+
+		logger.Infow("SOCKS_PROXY specified, SOCKS proxy will be used as a fallback for outgoing connection")
+		// start the main dial attempt.
+		primaryCtx, primaryCancel := context.WithCancel(ctx)
+		defer primaryCancel()
+		wg.Add(1)
+		primaryDial := func(ctx context.Context) (net.Conn, error) {
+			var zeroDialer net.Dialer
+			return zeroDialer.DialContext(ctx, network, addr)
+		}
+		go dialer(primaryCtx, primaryDial, true)
+
+		// wait a small amount before starting the fallback dial (to prioritize the primary connection method).
+		fallbackTimer := time.NewTimer(time.Second)
+		defer fallbackTimer.Stop()
+
+		// fallbackCtx is defined here because this fails `go vet` otherwise. The intent is for fallbackCancel
+		// to be called as this function exits, which will cancel the ongoing SOCKS proxy if it is still running.
+		fallbackCtx, fallbackCancel := context.WithCancel(ctx)
+		defer fallbackCancel()
+
+		// a for loop is used here so that we wait on both results and the fallback timer at the same time.
+		// if the timer expires, we should start the fallback dial and then wait for results.
+		// if the results channel receives a message, the message should be processed and either return
+		// or continue waiting (and reset the timer if it hasn't already expired).
+		for {
+			select {
+			case <-fallbackTimer.C:
+				wg.Add(1)
+				fallbackDial := func(ctx context.Context) (net.Conn, error) {
+					return socksProxyDialContext(ctx, network, proxyAddr, addr)
+				}
+				go dialer(fallbackCtx, fallbackDial, false)
+			case res := <-results:
+				if res.error == nil {
+					if res.primary {
+						logger.Infow("connected without SOCKS proxy")
```

Comment:
```suggestion
						logger.Infow("connected with ethernet/wifi")
```
[nit] Just to be more clear about the two methods of connection here.